File: 01/puzzle_1.py
from itertools import starmap


def solve(filepath):
  with open(filepath, "r") as file:
    left, right = zip(*(map(int, line.split()) for line in file))

  return sum(starmap(lambda a, b: abs(a - b), zip(sorted(left), sorted(right))))


result = solve("input.txt")

print(result)

File: 02/puzzle1.jl
using Pipe: @pipe

function is_safe(report)
  asc = nothing
  for (a, b) in zip(report[1:end-1], report[2:end])
    increasing = b > a
    diff = abs(a - b)

    if diff > 3 || diff == 0
      return false
    end

    if asc == nothing
      asc = increasing
      continue
    end

    if asc != increasing
      return false
    end
  end

  return true
end


function solve()
  return @pipe (
    open("input.txt", "r")
    |> readlines
    |> map(line -> parse.(Int, split(line)), _)
    |> filter(report -> is_safe(report), _)
    |> length
  )
end

println(solve())

